Update V 0.3

JetsadaBinaibrohim-Doppiotech opened the pull request:
1.ย้าย Resource ทั้งหมดที่ประกาศไล่ตามตัว และนำไปไว้ในไฟล์ set.robot ทั้งหมด 2.เพิ่ม registered_keyword.robot  เข้าไปเพิ่มจะได้เทสแบบไม่ต้อง register ใหม่ซ้ำๆ

JetsadaBinaibrohim-Doppiotech committed:
Update/Edit

1.ย้าย Resource ทั้งหมดที่ประกาศไล่ตามตัว และนำไปไว้ในไฟล์ set.robot ทั้งหมด
2.เพิ่ม registered_keyword.robot  เข้าไปเพิ่มจะได้เทสแบบไม่ต้อง register ใหม่ซ้ำๆ

JetsadaBinaibrohim-Doppiotech committed:
Update data

เปลี่ยน test 1 test มีหลาย test step

JetsadaBinaibrohim-Doppiotech committed:
Update

ผมลบและอัปเดตไฟล์ที่คิดว่าถูกต้องที่สุดเข้ามาแทน

JetsadaBinaibrohim-Doppiotech committed:
Update 0.3

Add README.txt files to notify lasted version  of assginment

massarinbdoppio commented on Doppee/.gitignore:
```diff
@@ -0,0 +1,8 @@
+*.gif
+*.bmp
+*.svg
+*.webp
```
น่าจะมี .log กับ .html ด้วยนะคะ

massarinbdoppio commented on Doppee/Testcase/tc.robot:
```diff
@@ -1,19 +1,11 @@
-Resource        ../Keywords/Feature/searchitem_keyword.robot
-Resource        ../Keywords/Feature/deliveryitem_keyword.robot
-Resource        ../Keywords/Feature/payment_keyword.robot
+Resource            ../Resource/set_up.robot
```
คิดว่า setup ของเราคือพวก import  file ป่ะคะ พวก library  variables resource ช่ายม่ะ
มันต้องชื่อ import.robot

massarinbdoppio commented on Doppee/Resource/set_up.robot:
```diff
@@ -0,0 +1,38 @@
+***Settings***
```
อ่าช่าย สิ่งนี้เราเรียกกันว่า import.robot ค่ะ แก้ชื่อหน่อย
กับมันไม่อยู่ใน test data ค่ะ มันควรอยู่ข้างนอกเลย เพราะมันไม่ได้ import แค่ testdata เท่านั้นอ่ะ

massarinbdoppio commented on Doppee/Testcase/log_data/geckodriver-1.log:
```diff
@@ -0,0 +1,12 @@
+1739249576006	geckodriver	INFO	Listening on 127.0.0.1:53177
```
ลบไฟล์นี้ทิ้งค่ะ ไม่เอาขึ้นมา git

massarinbdoppio commented on Doppee/Testcase/tc.robot:
```diff
@@ -1,19 +1,11 @@
-    [Tags]      Payment
-    Payment Product Method
\ No newline at end of file
+TC01 - Open Page Until Check Order
```
เฮ้ยยยย ชื่อเทสเคส ไม่ถูกต้องตามหลักการที่เรียนมาเลย ไปตั้งชื่อเทสเคสมาใหม่
มันคือ Test objective อ่ะ

massarinbdoppio commented on Doppee/Testcase/tc.robot:
```diff
@@ -1,19 +1,11 @@
-    Payment Product Method
\ No newline at end of file
+TC01 - Open Page Until Check Order
+    Open Doppee Website
+    Registeration Account
+    Searching Product
+    Delivery Infomation Product
+    Payment Method
+    Check Product Order Create
```
step ได้อยู่ แต่ว่าไม่มี reference เลยคับ ไปใส่มาให้ถูก

massarinbdoppio commented on Doppee/Testcase/tc.robot:
```diff
@@ -1,19 +1,11 @@
\ No newline at end of file
+TC01 - Open Page Until Check Order
+    Open Doppee Website
+    Registeration Account
```
แต่ละ step  เราไม่ส่งตัวแปรเลยเหรอ 
เช่น search product  อะไร  register ด้วย  ค่าอะไร อันนี้ถาม idea ค่ะว่า ทำไมไม่มีส่ง

massarinbdoppio commented on Doppee/Resource/Testdata/data.yaml:
```diff
@@ -0,0 +1,35 @@
+url_doppee    : "https://training-platform.doppio-tech.com"
+type_browser  : "gc"
+env_config    :
+  uat         : "https://training-platform.doppio-tech.com"
```
วิธีนี้ ไม่ตอบโจทย์ที่เค้าสั่งมาค่ะ

massarinbdoppio commented on Doppee/Resource/Locator/variables_path/register_path.robot:
```diff
@@ -0,0 +1,8 @@
+***Variables***
```
ใต้ locator เราไม่เรียกพวกนี้ว่า variables_path แล้วค่ะ ไม่จำเป็นต้องมี folder อีกชั้น
กับชื่อไฟล์ เราก็เรียกว่า locator ค่ะ ไม่เรียก path
สามารถเป็นแบบนี้ได้
Locator >  payment_locator.robot

แก้ทุกไฟล์ของ locator นะ

massarinbdoppio commented on Doppee/Resource/Locator/variables_path/checkorder_path.robot:
```diff
@@ -0,0 +1,5 @@
+***Variables***
+${payment_continue}     //button/span[text()='Continue Shopping']
+${product_order}        //tr[@class='ant-table-row ant-table-row-level-0']
+${status_id}            //td[@class='ant-table-cell'][text()!='Pending payment']
+${status_order}         //td[@class='ant-table-cell'][text()='Pending payment']
```
2 ค่านี้ ทำไม xpath มันดูไม่สอดคล้องเลย อธิบายหน่อย

JetsadaBinaibrohim-Doppiotech replied:
ตรงนี้ผมไปดูมาหลังจากกดชำระเงินแล้ว กดไปที่ Green User Icon  แล้วเลื่อนลงมาดู My order ผมกดคลิกขวา Inspect เพื่อจะหา xpath ของ ตัวเลข ID กับ ตัวหนังสือ Pending payment ว่ามีขึ้นบนหน้าจอไหม

ซึ่งทั่งสองตัวนี้ใช้   //td[@class='ant-table-cell'] เหมือนกันแตกต่างกันตรงที่ text ข้างในอย่างเดียว
เนื่องจาก Order ID เป็นค่าไม่คงที่เปลี่ยนได้ตลอด ผมเลยใช่เป็น //td[@class='ant-table-cell'][text()='Pending payment']
ที่เป็นค่าคงที่แน่ๆ  เพื่อเช็คว่ามีตัวหนังสือ Pending payment ขึ้นไหม และ //td[@class='ant-table-cell'][text()!='Pending payment'] เพื่อเช็คว่ามีตัวเลขขึ้นไหมครับ

massarinbdoppio commented on Doppee/Resource/Locator/commom_locator.robot:
```diff
@@ -0,0 +1,2 @@
+***Variables***
```
แก้ชื่อไฟล์นี้ว่า header_locator ได้ค่ะ 
ด้านบนนั้น มันไม่ common หนะ มันเรียกว่า header ได้ หรือ header_menu ไรงี้ก็ได้
เอาเป็น header_menu_locator ละกันค่ะ

massarinbdoppio commented on Doppee/Keyword/Page/searchproduct_page.robot:
```diff
@@ -1,16 +1,14 @@
+Search Product On Search Bar
+    Wait Until Element Is Visible       ${place_holder}         ${timeout}
+    Input Text                          ${input_product}        ${product}
+Click Search Icon Or Hit Enter Keyboard
```
เว้น 1 บรรทัด ทุก keyword

massarinbdoppio commented on Doppee/Keyword/Page/searchproduct_page.robot:
```diff
@@ -1,16 +1,14 @@
-    Wait Until Element Is Visible       ${product_modal}        ${timeout}
+***Keywords***
+Search Product On Search Bar
+    Wait Until Element Is Visible       ${place_holder}         ${timeout}
```
ไม่มี reference คับ เติมมาให้หมดค่ะ ทุกไฟล์ 
**SeleniumLibrary**.Wait Until Element Is Visible

massarinbdoppio commented on Doppee/Keyword/Page/register_page.robot:
```diff
@@ -0,0 +1,25 @@
+***Keywords***
+Go To Registeration
```
สะกดผิด

massarinbdoppio replied:
Go to register page ควรอยู่ใน feature level ค่ะ เพราะมันไม่ช่าย action ของหน้า register ถูกม่ะ 
มันกดอะไรตั้งหลายอย่างที่อยู่หน้าอื่นๆ มาจนถึงหน้า register ต่างหาก

massarinbdoppio commented on Doppee/Keyword/Page/register_page.robot:
```diff
@@ -0,0 +1,25 @@
+    ${email}=       Set Variable        ${username}${random_number}${domain}
+    RETURN          ${email}
+
+Input Username Password And Comfirm
```
level page 
1 action = 1 keyword

keyword นี้ของเรา คือ 3 action แล้ว
1. input username
2. input password
3. input confirm password

massarinbdoppio commented on Doppee/Keyword/Page/payment_page.robot:
```diff
@@ -0,0 +1,18 @@
+***Keywords***
+Go To Payment Method
```
level feature เหตุผลเหมือนของ register page

massarinbdoppio replied:
Resolved แบบไม่ได้ย้ายไป feature 
มีเหตุผลว่ายังไงเอ่ย

JetsadaBinaibrohim-Doppiotech replied:
![Screenshot 2025-02-19 144228](https://github.com/user-attachments/assets/3deea2eb-34dd-4fb1-9187-9ec7429f997d)
อันนี้ผมย้ายแล้วนะครับ ใช่แบบนี้ไหมครับ

massarinbdoppio replied:
พี่น่าหมายถึงย้ายทั้ง kw ไป feature คับ
ไม่เป็นไรค่ะ ค้างไว้ที่นี่ก่อนได้

massarinbdoppio commented on Doppee/Keyword/Page/payment_page.robot:
```diff
@@ -0,0 +1,18 @@
+    SeleniumLibrary.Click Element                       ${pay_button}
+    SeleniumLibrary.Page Should Contain                 ${method}
+    SeleniumLibrary.Click Element                       ${next_button}  
+Input Payment Details
```
เว้น 1 บรรทัด ทุกที่ ทุกไฟล์ ทำมาใหม่นะคะ ไม่ comment ให้ทุก keyword

massarinbdoppio replied:
input payment detail
มีเหตุผลที่ไม่แยก 1 action = 1 kw ไหม 
แบบจิลมองว่า ทำไมไม่ต้องแยก หรือลืมคิด

massarinbdoppio commented on Doppee/Keyword/Page/deliveryinfo_page.robot:
```diff
@@ -0,0 +1,16 @@
+***Keywords***
+Go To Delivery Infomation
```
level feature เหตุผลเหมือนของ register page

massarinbdoppio commented on Doppee/Keyword/Page/deliveryinfo_page.robot:
```diff
@@ -0,0 +1,16 @@
+Go To Delivery Infomation
+    SeleniumLibrary.Click Element                       ${cart_icon}
+    SeleniumLibrary.Wait Until Element Is Visible       ${form_info}        ${timeout}
+Input User Infomation Address
```
มีเหตุผลที่ไม่แยก 1 action = 1 kw ไหม
แบบจิลมองว่า ทำไมไม่ต้องแยก หรือลืมคิด

massarinbdoppio commented on Doppee/Keyword/Page/checkorder_page.robot:
```diff
@@ -0,0 +1,9 @@
+***Keywords***
+Go To Doppee Website
```
ชื่อไม่สื่อค่ะ แก้หน่อย

massarinbdoppio commented on Doppee/Keyword/Page/checkorder_page.robot:
```diff
@@ -0,0 +1,9 @@
+***Keywords***
+Go To Doppee Website
+    SeleniumLibrary.Click Element                       ${payment_continue} 
+Click Green User Icon
```
green user icon ไม่ได้อยู่หน้า check order page อยู่ใน header_menu_page อะไรแบบนี้มากกว่า

massarinbdoppio commented on Doppee/Keyword/Page/checkorder_page.robot:
```diff
@@ -0,0 +1,9 @@
+Click Green User Icon
+    SeleniumLibrary.Click Element                       ${user_icon}
+    SeleniumLibrary.Wait Until Element Is Visible       ${product_order}        ${timeout}
+Check Order Status 
```
check order status พี่น่าคิดว่าควรมีการรับ argument เป็น order id ที่ต้องการจะ check เข้ามานะคะ 
การ fix  ${status_id}  คือ เช็คของ order ไหนอ่ะ ไม่บอก ไม่ชัดเจนค่ะ

kw นี้ต้องแก้ logic ค่ะ เช็คแค่ not be empty ไม่พอค่ะ

massarinbdoppio commented on Doppee/Keyword/Feature/searchproduct_keyword.robot:
```diff
@@ -0,0 +1,6 @@
+***Keywords***
+Searching Product
+    Search Product On Search Bar
+    Click Search Icon Or Hit Enter Keyboard
```
ตกลง kw นี้ click หรือ hit enter

massarinbdoppio commented on Doppee/Keyword/Feature/register_keyword.robot:
```diff
@@ -0,0 +1,7 @@
+***Keywords***
+Registeration Account
```
สะกดผิดนะคิดว่า

massarinbdoppio commented on Doppee/Keyword/Feature/register_keyword.robot:
```diff
@@ -0,0 +1,7 @@
+***Keywords***
+Registeration Account
+    Go To Registeration
+    Generate Username                       ${username}         ${domain}
```
จิงๆ ควรให้คนที่เรียน kw นี้เป็นคนส่ง username เข้ามาหรือป่าวคะ 
อันนี้ถาม algorithm เราแหละว่าเราออกแบบยังไงนะ

massarinbdoppio commented on Doppee/Keyword/Feature/payment_keyword.robot:
```diff
@@ -0,0 +1,7 @@
+    Input Payment Details
+    Click Confirm Payment
+    Pop-up Confirm Payment Success
+    Order Information Created
```
ชื่อ kw ค่ะ แก้หน่อย 
Verify Order Information Created หรือไม่ก็ Check Order Information Created

massarinbdoppio commented on Doppee/Keyword/Feature/deliveryinfo_keyword.robot:
```diff
@@ -0,0 +1,5 @@
+Delivery Infomation Product
+    Go To Delivery Infomation
+    Input User Infomation Address
+    Delivery Infomation Match with Testdata
```
Delivery Infomation Match with Testdata - เช็คทำไมนะคะ เพราะเราก็เพิ่งกรอกไปนี่หน่า ถามนะคะว่าทำไมถึงต้องเช็คอีกอ่ะ คิดยังไง

JetsadaBinaibrohim-Doppiotech replied:
#อันนี้ผมเช็คว่าในหน้าเพจนี้ข้อมูลตรงกับตัว test data ที่เราเอามา test ว่าตรงกันไหม ผมเลยคิดว่าจำเป็นต้องใส่มาครับ

massarinbdoppio commented on Doppee/Keyword/Feature/checkorder_keyword.robot:
```diff
@@ -0,0 +1,5 @@
+***Keywords***
+Check Product Order Create
+    Go To Doppee Website
```
หน้าแรกของ Doppee Website เราเรียก หน้า Home page ก็ได้ค่ะ

massarinbdoppio reviewed (changes requested):
อ่ะแก้มาค่ะ

JetsadaBinaibrohim-Doppiotech committed:
Update V 0.4

- เพิ่ม *.log , *.html ใน github
- แก้ไฟล์ set_up.robot เป็น import.robot และย้ายไฟล์ออกมานอกโฟลเดอร์ Resource
- ลบไฟล์ geckodriver-1.log
- แก้ชื่อ Test case และเพิ่ม Tags
- แก้ตัวหนังสือพิมพ์ผิด Registration
- แก้ Keyword ใน payment_keyword.robot ให้เข้าใจง่าย
- เปลี่ยนชื่อหน้าแรกเป็น Home page
- เพิ่ม comment ในส่วนที่พี่รีน่าถาม 2 comment

massarinbdoppio commented on Doppee/Testcase/tc.robot:
```diff
@@ -1,19 +1,12 @@
\ No newline at end of file
+TC01 - Test Doppee Website Functional
+    [Tags] Website
+    Open Doppee Website
```
ขาด reference

massarinbdoppio replied:
ขาด references

register_keyword.Registration Account

massarinbdoppio commented on Doppee/Keyword/Feature/register_keyword.robot:
```diff
@@ -1,7 +1,9 @@
+Registration Account
+    Go To Registration
     Generate Username                       ${username}         ${domain}
+#ตัวนี้เป็น Arguments username ที่รับมาจาก register_page.robot ที่ผมได้มีการ random เลข 1000-9999 ไว้หลัง แล้วเก็บไว้ใน ${email}
```
ลบทิ้ง

massarinbdoppio commented on Doppee/Keyword/Page/checkorder_page.robot:
```diff
@@ -1,5 +1,5 @@
 ***Keywords***
-Go To Doppee Website
+Go To Home Page
```
ไม่อยุ่ในนี้ 
อ่ะ อยู่ใน header_menu_page อะไรแบบนี้มากกว่า

massarinbdoppio commented on Doppee/Resource/Testdata/data.yaml:
```diff
@@ -0,0 +1,35 @@
+price         : "49980"
+total         : "Total : 49980 THB"
+qty           : "1"
+
```
product_stock_0: zoppee

massarinbdoppio commented on Doppee/Keyword/Page/searchproduct_page.robot:
```diff
@@ -1,16 +1,14 @@
-    Click Element                       ${product_add}
-    Wait Until Element Is Visible       ${product_modal}        ${timeout}
+***Keywords***
+Search Product On Search Bar
```
แก้ให้รองรับการ search ด้วย product ชื่ออะไรก็ได้ตามที่ test case ต้องการ

massarinbdoppio commented on Doppee/Testcase/tc.robot:
```diff
@@ -1,10 +1,11 @@
 
 ***Test Cases***
-TC01 - Open Page Until Check Order
+TC01 - Test Doppee Website Functional
```
ชื่อเทสเคสยังไม่ผ่าน

JetsadaBinaibrohim-Doppiotech committed:
Update assignment V0.5

    แก้ path เป็น locator และเปลี่ยนชื่อ โฟลเดอร์ commom_locator เป็น header_menu_locator
    แก้ชื่อ test case
    เพิ่ม reference ให้กับทุก  keyword
    เว้นบรรทัดแต่ละ keyword ให้อ่านง่ายขึ้น
    เพิ่มไฟล์ header_menu_locator และ header_menu_page
    ลบไฟล์ uat  sit ที่ไม่ได้ใช้งาน
    แยก 1 action = 1 keyword ที่พี่รีน่าคอมเมนต์
    แยก ส่วนของ level feature ที่อยู่ใน page ไปไว้ในส่วน feature

JetsadaBinaibrohim-Doppiotech committed:
Update assignment 0.5.1

ยังไม่ได้เพิ่ม Logic ของส่วน Order ID ครับ

massarinbdoppio commented on Doppee/Resource/Locator/searchproduct_locator.robot:
```diff
@@ -0,0 +1,9 @@
+${input_product}        //input[@type='text']
+${card_env}             //div[@class='product-card-container']
+${search_icon}          //span/button[@type='button']
+${product_card}         //div[@class='ant-card-cover']/img[@src='https://t4.ftcdn.net/jpg/03/01/83/01/360_F_301830112_dSHXcFf9OHOj6HBpaDsAa1gAtz3hwheP.jpg']
```
จิงๆ ไม่ควร fix แบบนี้นะคะ

massarinbdoppio commented on Doppee/Resource/Locator/searchproduct_locator.robot:
```diff
@@ -0,0 +1,9 @@
+${input_product}        //input[@type='text']
+${card_env}             //div[@class='product-card-container']
+${search_icon}          //span/button[@type='button']
+${product_card}         //div[@class='ant-card-cover']/img[@src='https://t4.ftcdn.net/jpg/03/01/83/01/360_F_301830112_dSHXcFf9OHOj6HBpaDsAa1gAtz3hwheP.jpg']
```
จิงๆ ไม่ควร fix แบบนี้นะคะ แต่รอบนี้ไม่ต้องแก้ได้ค่ะ
แต่ให้ตอบมาหน่อยว่า คิดว่าทำไมถึงไม่ควร fix ค่ะ มันจะเกิดปัญหาอะไร

JetsadaBinaibrohim-Doppiotech replied:
ตรงนี้ผมคิดว่าควรจะ fix อ่ะครับพี่รีน่า เพราะว่ามันเป็น path ของการทำ action ถ้ามีตัวไหนไม่ได้ใส่หรือใส่ผิดมันก็จะหยุดทำงานแค่เท่าที่ไปต่อได้ครับ

massarinbdoppio commented on Doppee/Resource/Locator/deliveryinfo_locator.robot:
```diff
@@ -0,0 +1,11 @@
+${address_info}         //input[@placeholder='Shipping address'  and @type='text']
+${phone_info}           //input[@placeholder='Phone number'  and @type='text']
+${email_info}           //input[@placeholder='Email'  and @type='text']
+${order_name}           //div[@class='product-detail-item' and text()='Coppee phone 1']
```
ทำไมถึง fix ค่าเป็น Coppee phone 1 
ไม่ต้องแก้ แต่ตอบคำถามมาว่า มันจะมีข้อจำกัดยังไง

JetsadaBinaibrohim-Doppiotech replied:
ค่าของ Coppee phone 1 ผมสร้างมาแล้วไม่ได้ถูกเอาไปใช้เลยครับ สร้างมาแล้วลืมลบตัวแปรออก เหตุผลที่สร้างมาเพื่อจะมาดูว่าค่าตรงกับที่เลือกไว้ไหมหลัง Search

massarinbdoppio commented on Doppee/Resource/Locator/deliveryinfo_locator.robot:
```diff
@@ -0,0 +1,11 @@
+${phone_info}           //input[@placeholder='Phone number'  and @type='text']
+${email_info}           //input[@placeholder='Email'  and @type='text']
+${order_name}           //div[@class='product-detail-item' and text()='Coppee phone 1']
+${order_price}          //div[@class='product-detail-item' and text()='49980']
```
ทำไมถึง fix ค่าเป็น 49980
ไม่ต้องแก้ แต่ตอบคำถามมาว่า มันจะมีข้อจำกัดยังไง

JetsadaBinaibrohim-Doppiotech replied:
เหตุผลเดียวกับ Coppee phone 1 เลยครับคือสร้างตัวแปรมาไม่ได้ใช้ ลืมลบค่าออก

massarinbdoppio commented on Doppee/Resource/Locator/checkorder_locator.robot:
```diff
@@ -0,0 +1,6 @@
+***Variables***
+${payment_continue}     //button/span[text()='Continue Shopping']
+${product_order}        //tr[@class='ant-table-row ant-table-row-level-0']
+${order_id}             //td[@class='ant-table-cell'][text()!='Pending payment']
+${status_id}            //td[@class='ant-table-cell'][text()='Pending payment']         
```
2 ค่านี้ อธิบายพี่น่าหน่อยว่า xpath มันคือเอาไว้เช็คอะไรนะคะ

JetsadaBinaibrohim-Doppiotech replied:
ในตัว Element ของ xpath ชื่อเหมือนกันทั้งสองตัวเลยครับแล้วผมต้องการเช็คว่า มีข้อความขึ้นหรือไม่
ผมเลยใช้ xpath เขียนไปว่า ค่านึงต้องมีตัวหนังสือที่ระบุและอีกค่าที่ไม่ใช่ตัวหนังสือระบุครับ

สรุปก็คือ ในหน้า My order มันจะมี  2 ค่า คือ  Order ID กับ Status ตัว ${order_id} ก็จะเจอค่าที่ไม่ใช่  Pending payment แล้วจะไป match กับ อีกค่าที่เป็นตัวหนังสือที่ระบุ เลข IDไว้ ส่วน ${status_id}  ก็หาตัวหนังสือตรงตาม Pending payment เลยครับ

massarinbdoppio commented on Doppee/Testcase/tc.robot:
```diff
@@ -1,19 +1,12 @@
+    searchproduct_keyword.Searching Product                                   
+    deliveryinfo_keyword.Delivery Infomation Product
+    payment_keyword.Payment Method
+    checkorder_keyword.Check Product Order Create
```
โอเค เขียนมาถูกทางแล้วค่ะ
สิ่งที่ทำให้ check order ยังไม่ถูกนัก เพราะว่า จิลไม่มีเลข order id ไปเช็คคับ 
แนะนำให้ get order id จากหน้า Payment Confirm หนะ ที่กดจ่ายเงินแล้วได้เลข order id
ทีนี้ก็ค่อยส่งค่า order id ไป check product order created ว่าที่หน้า pending payment มันมี order id ตรงกันนะ

JetsadaBinaibrohim-Doppiotech replied:
อันนี้ผมหา xpath ของ order id หลังกดชำระเงินแล้วขึ้นหน้า payment complete แล้วไม่เจอครับ ผมมองภาพออกแต่เขียน logic ไม่ได้ครับ ลองดึงค่ามาใช้แล้วแต่ขึ้น error ตลอดเลย

massarinbdoppio reviewed (approved):
Approve ให้ผ่าน เนื่องจากจะมี assignment ต่อเนื่องที่น่าจะเข้ามาทำให้ต้องมีการปรับแก้เพิ่มเติมต่อ
